File: leetcode/C++/network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<int>> p(n+1), w(n+1);
        vector<int> visit(n+1,0);
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        int sol=0, node, val, next_node, next_val;
        for(int i=0; i<times.size(); ++i){
            p[times[i][0]].push_back(times[i][1]);
            w[times[i][0]].push_back(times[i][2]);
        } 
        pq.push({0,k});
        while(!pq.empty() && n){
            node=pq.top().second; val=pq.top().first; pq.pop();
            if(visit[node]) continue;
            visit[node]=1; --n;
            if(sol<val) sol=val;
            for(int i=0; i<p[node].size()&&n; ++i){
                if(visit[p[node][i]]) continue;
                next_node=p[node][i]; next_val=val+w[node][i];
                pq.push({next_val,next_node});
            }
        }
        return (n==0)? sol : -1;
    }
};

File: leetcode/Python/network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        m, pq, sol, visit = [[] for i in range(len(times)+2)],[], -1, [0 for i in range(len(times)+2)]
        for u,v,w in times:
            m[u].append([v,w])
        heapq.heappush(pq,[0,k])
        while pq and n:
            w, node = heapq.heappop(pq)
            if visit[node]>0:
                continue
            visit[node]=1
            n-=1
            sol=max(sol,w)
            for ne, nw in m[node]:
                if visit[ne]>0:
                    continue
                heapq.heappush(pq,[nw+w,ne])
        return sol if n==0 else -1